src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs

```diff
@@ -267,23 +272,51 @@ private void EnsureServer()
             }
         }
 
-        public void Dispose()
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
+            if (_stopped)
+            {
+                return;
+            }
+            _stopped = true;
+
             _logger?.Shutdown();
 
+            if (!cancellationToken.CanBeCanceled)
+            {
+                cancellationToken = new CancellationTokenSource(Options.ShutdownTimeout).Token;
+            }
+
             // Fire IApplicationLifetime.Stopping
             _applicationLifetime?.StopApplication();
 
+            try
+            {
+                await Server?.StopAsync(cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                _logger?.ServerShutdownException(ex);
```

Comment:
While Dispose shouldn't throw in a perfect world, it's the only way to communicate errors back to Program.Main for any app that doesn't call the new Stop API. I assume there will be plenty of apps that only call Dispose and not Stop considering Dispose calls Stop internally.